quote.py

```diff
@@ -3,48 +3,102 @@
 import random
 from halibot import HalModule
 
+sr = random.SystemRandom()
+
 class Quote(HalModule):
 
-	quotes = []
+	# Configuration defaults
+	remove_votes = 3
+	max_history = 10
 
 	def init(self):
-		self.path = self.config['quotes-path']
+		self.quotes = []
+		self.path = self.config.get('quotes-path')
+		self.remove_votes = self.config.get('remote-votes', self.remove_votes)
+		self.max_history = self.config.get('max-history', self.max_history)
 		print('using quotes path:', self.path)
 
-		with open(self.path, 'r') as f:
-			self.quotes = [x.rstrip() for x in f.readlines()]
+		if self.path is not None:
+			try:
+				with open(self.path, 'r') as f:
+					self.quotes.extend([x.rstrip() for x in f if x.rstrip()])
+			except IOError, OSError:
+				pass
+
+		self.rand_quotes = {}  # pattern -> shuffled lists
+		self.rem_vote_map = {}  # quote -> set(usernames)
 
 	def quote(self, pattern=""):
-		if len(pattern) > 0:
-			try:
-				expr = re.compile(pattern)
-			except re.error as e:
-				return 'Invalid pattern: ' + str(e)
-				
-			ls = [q for q in self.quotes if re.search(expr, q)]
+		if pattern in self.rand_quotes:
+			ls = self.rand_quotes[pattern]
+			quote = ls.pop()
+			if not ls:
+				del self.rand_quotes[pattern]
 		else:
-			ls = self.quotes
+			if len(pattern) > 0:
+				try:
+					expr = re.compile(pattern)
+				except re.error as e:
+					return 'Invalid pattern: %s' % (e,)
```

Comment:
For a second, I wasn't sure if this was Python 3, so I was being as close to Python 2 spec as possible :P 

If you don't mind me dropping Py2 compat, I'll gladly do that right now :D